### PR TITLE
Do not allow addition of duplicate team members on an escalation

### PR DESCRIPTION
**Pull request by preeti-preeti**
@pcon please review this

### CODE REVIEW

File: src/classes/EscalationTeamTrigger.cls [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+		Map<Id, Set<Id>> escalToMembers = new Map<Id, Set<Id>>();
+		Map<Id, Set<String>> escalToEmails = new Map<Id, Set<String>>();
+
+		for (rh_escal__EscalationTeam__c team: EscalationTeamUtils.getEscalationTeamMembers(escalationIds)) {
+			if (!escalToMembers.containsKey(team.rh_escal__Escalation__c)) {
+				escalToMembers.put(team.rh_escal__Escalation__c, new Set<Id>());
+				escalToEmails.put(team.rh_escal__Escalation__c, new Set<String>());
+			}
+
+			if (team.rh_escal__Contact__c != null) {
```
Comment (pcon) [inline src/classes/EscalationTeamTrigger.cls]: We can save some time here by just saying

```java
    escalToMembers.get(team.rh_escal__Escalation__c).add(team.rh_escal__Contact__c);
    escalToMembers.get(team.rh_escal__Escalation__c).add(team.rh_escal__User__c);
    escalToEmail.get(team.rh_escal__Escalation__c).add(team.rh_escal__Email__c);
}

escalToMembers.remove(null);
escalToEmails.remove(null);

File: src/classes/EscalationTeamTrigger.cls [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+		if (escalationIds.isEmpty()) {
+			return;
+		}
+
+		Map<Id, Set<Id>> escalToMembers = new Map<Id, Set<Id>>();
+		Map<Id, Set<String>> escalToEmails = new Map<Id, Set<String>>();
+
+		for (rh_escal__EscalationTeam__c team: EscalationTeamUtils.getEscalationTeamMembers(escalationIds)) {
+			if (!escalToMembers.containsKey(team.rh_escal__Escalation__c)) {
```
Comment (pcon) [inline src/classes/EscalationTeamTrigger.cls]: Remove this and build up the map outside of the loop.  This will mean we don't have to do a null check below

```java
for (Id escalationId : escalationIds) {
    escalToMembers.put(escalationId, new Set<Id>());
    escalToEmail.put(escalationId, new Set<Id>());
}
```

File: src/classes/EscalationTeamTrigger.cls [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		}
+
+		for (rh_escal__EscalationTeam__c team: this.newObjs) {
+			Set<Id> members = escalToMembers.get(team.rh_escal__Escalation__c);
+			Set<String> emails = escalToEmails.get(team.rh_escal__Escalation__c);
+
+			Boolean isContactNull = team.rh_escal__Contact__c == null;
+			Boolean isUserNull = team.rh_escal__User__c == null;
+			Boolean isEmailNull = team.rh_escal__Email__c == null;
+			Boolean isMembersNull = members == null;
```
Comment (pcon) [inline src/classes/EscalationTeamTrigger.cls]: With the update above, we should be able to get rid of `isMembersNull` and `isEmailsNull` here and in the checks on lines 111-115

File: src/classes/EscalationTeamUtils.cls [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+			select rh_escal__Contact__c,
+				rh_escal__Email__c,
+				rh_escal__Escalation__c,
+				rh_escal__Internal__c,
+				rh_escal__Type__c,
+				rh_escal__User__c
+			from rh_escal__EscalationTeam__c
+			where rh_escal__Escalation__c in :escalIds
+		];
+
```
Comment (pcon) [inline src/classes/EscalationTeamUtils.cls]: remove blank line

File: src/classes/TestUtils.cls [hunk 5]
```diff
@@ -141,4 +178,138 @@ global class TestUtils {
...
+	/**
+	* Fetches a User with an Id
+	*
+	* @param u A userId
+	* @return A User
+	*/
+	global static User fetchUser(Id userId) {
+		return [
+			select
+				Alias,
```
Comment (pcon) [inline src/classes/TestUtils.cls]: move up to previous line and remove any fields you don't need for the test.

File: src/classes/TestUtils.cls [hunk 6]
```diff
@@ -141,4 +178,138 @@ global class TestUtils {
...
+		return [
+			select Name,
+				rh_escal__Contact__c,
+				rh_escal__Email__c,
+				rh_escal__Escalation__c,
+				rh_escal__Internal__c,
+				rh_escal__Type__c,
+				rh_escal__User__c
+			from rh_escal__EscalationTeam__c
+			where Id =:escalTeam.id
```
Comment (pcon) [inline src/classes/TestUtils.cls]: space between `=` and `:escalTeam.id`

Also capitalize Id

File: src/classes/TestUtils.cls [hunk 7]
```diff
@@ -141,4 +178,138 @@ global class TestUtils {
...
+		Profile adminProfile = PROFILE_NAME_MAP.get('System Administrator');
+
+		User admin = [
+			select Name,
+				ProfileId,
+				Profile.Name
+			from User
+			where ProfileId = :adminProfile.Id and
+				isActive = true and
+				LocaleSidKey = 'en_US'
```
Comment (pcon) [inline src/classes/TestUtils.cls]: Remove this requirement

File: src/classes/EscalationTeamTrigger_Test.cls [hunk 8]
```diff
@@ -0,0 +1,480 @@
+/**
+* Test for the escalation team trigger
+*
+* @author Preeti (ppreeti@redhat.com)
+*/
+
+@IsTest
+public with sharing class EscalationTeamTrigger_Test {
```
Comment (pcon) [inline src/classes/EscalationTeamTrigger_Test.cls]: Can you add some tests that show this working?  Like adding two team members that are not duplicates?  And updating a team member to something that is not a duplicate.

File: src/classes/EscalationTeamUtils_Test.cls [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+		List<rh_escal__EscalationTeam__c> results = EscalationTeamUtils.getEscalationTeamMembers(escalationIds);
+
+		Test.stopTest();
+
+		System.assertEquals(3, results.size(), 'Did not get the expected number of results back');
+		System.assertEquals(expectedResults.size(), results.size(), 'Did not get the expected number of results back');
+
+		for (rh_escal__EscalationTeam__c escalTeam : results)
+		{
```
Comment (pcon) [inline src/classes/EscalationTeamUtils_Test.cls]: put this on the previous line

File: src/classes/TestUtils.cls (hunk 5 above)
Comment (preeti-preeti) [inline src/classes/TestUtils.cls]: I am not using this method, I just added this for future use. I am removing this complete method.

File: src/classes/EscalationTeamTrigger_Test.cls (hunk 8 above)
Comment (preeti-preeti) [inline src/classes/EscalationTeamTrigger_Test.cls]: setUniqueContactTypeMember_onAnEscalationTeamUpdateTest and other methods with such functionality are present in test class

Comment (pcon) [inline src/classes/EscalationTeamTrigger_Test.cls]: Ok, thanks.  Just missed it.

File: src/classes/EscalationTeamTrigger.cls [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+			escalToMembers.put(escalationId, new Set<Id>());
+			escalToEmails.put(escalationId, new Set<String>());
+		}
+
+		for (rh_escal__EscalationTeam__c team: EscalationTeamUtils.getEscalationTeamMembers(escalationIds)) {
+			escalToMembers.get(team.rh_escal__Escalation__c).add(team.rh_escal__Contact__c);
+			escalToMembers.get(team.rh_escal__Escalation__c).add(team.rh_escal__User__c);
+			escalToEmails.get(team.rh_escal__Escalation__c).add(team.rh_escal__Email__c);
+
+			escalToMembers.remove(null);
```
Comment (pcon) [inline src/classes/EscalationTeamTrigger.cls]: Move these two outside the for loop.  Sorry that I wasn't clearer earlier

File: src/classes/EscalationTeamTrigger.cls [hunk 11]
```diff
@@ -86,13 +86,13 @@ public with sharing class EscalationTeamTrigger extends BaseTrigger {
 			escalToEmails.put(escalationId, new Set<String>());
 		}
 
+		escalToMembers.remove(null);
```
Comment (pcon) [inline src/classes/EscalationTeamTrigger.cls]: This should be done after the for loop.

Comment (preeti-preeti) [inline src/classes/EscalationTeamTrigger.cls]: sorry my bad, didn't payed much attention in hurry
